### PR TITLE
Refactor storage account controller to reduce ARM API calls

### PR DESCRIPTION
**Pull request by tsatam**
### Which issue this PR addresses:

Fixes [ARO-4632](https://issues.redhat.com/browse/ARO-4632)

### What this PR does / why we need it:

Reduces ARM API calls in storageaccounts controller through the following changes:
- Handle HTTP 429 responses from Azure API calls by requeuing the controller after the specified "Retry-After" header duration (rather than immediately)
- Store the "last reconciled" time and subnets/storageaccounts, to avoid performing further reconciliation if the parameters have not changed

### Test plan for issue:

- Unit tests have been refactored to test the entire controller, not just the manager
- Looking for ideas on how to reproduce ARM rate limiting to test that functionality (e2e or more likely manual tests)

### Is there any documentation that needs to be updated for this PR?

No (no expected behavioral changes)

### CODE REVIEW

**Comment by tsatam** [comment]
Putting temp hold for now, as #3280 should merge first and this PR rebased on those changes, if we even want to ship this at all (#3280 may be sufficient to mitigate the API call volume that this PR seeks to reduce)

File: pkg/operator/deploy/deploy.go [hunk 1]
```diff
@@ -212,6 +212,13 @@ func (o *operator) resources() ([]kruntime.Object, error) {
 		serviceSubnets = append(serviceSubnets, "/subscriptions/"+o.env.SubscriptionID()+"/resourceGroups/"+o.env.GatewayResourceGroup()+"/providers/Microsoft.Network/virtualNetworks/gateway-vnet/subnets/gateway-subnet")
 	}
 
+	// this service subnet is added during install time if cluster is installed via Hive
+	if o.oc.Properties.HiveProfile.CreatedByHive {
+		// TODO: add AKS shard support
+		hiveShard := 1
+		serviceSubnets = append(serviceSubnets, fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Network/virtualNetworks/aks-net/subnets/PodSubnet-%03d", o.env.SubscriptionID(), o.env.ResourceGroup(), hiveShard))
+	}
```
Comment (SudoBrendan) [inline pkg/operator/deploy/deploy.go:220]: This change feels out-of-scope for this work, was it included by mistake or is it somehow reducing API calls?

**Review by SudoBrendan — COMMENTED** [review]
NOT TO BE READ BEFORE MONDAY, 11/27, OR THE TURKEY OF THANKSGIVING PAST _WILL_ HAUNT YOU :turkey: :ghost: 

A few nits, questions, and some changes I think might have been included by mistake, or are perhaps reflective of the rebase that needs to happen (looks like dep was merged, so that's good to go now).

File: pkg/operator/controllers/storageaccounts/storageaccount_controller.go [hunk 2]
```diff
@@ -75,40 +76,165 @@ func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.
...
+	if err != nil {
+		if retryAfter, ok := errIsRateLimited(err); ok {
+			return reconcile.Result{RequeueAfter: retryAfter}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	storageAccounts, err := r.getStorageAccounts(ctx, instance)
+	if err != nil {
+		return reconcile.Result{}, err
```
[the remark this answers is not in the document]
Comment (tsatam) [reply pkg/operator/controllers/storageaccounts/storageaccount_controller.go:114]: This is a case of poor naming on the function. The "storage accounts" retrieved here are just their names, derived from in-cluster resources. No Azure API calls are being made here. I'll update this to something like `getStorageAccountNames`.

File: pkg/operator/controllers/storageaccounts/storageaccount_controller.go [hunk 3]
```diff
@@ -75,40 +76,165 @@ func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.
...
-		client:      r.client,
-		kubeSubnets: subnet.NewKubeManager(r.client, resource.SubscriptionID),
-		subnets:     subnet.NewManager(&azEnv, resource.SubscriptionID, authorizer),
-		storage:     storage.NewAccountsClient(&azEnv, resource.SubscriptionID, authorizer),
+			return retryAfter, true
+		}
 	}
 
-	return reconcile.Result{}, manager.reconcileAccounts(ctx)
+	return 0, false
```
[the remark this answers is not in the document]
Comment (tsatam) [reply pkg/operator/controllers/storageaccounts/storageaccount_controller.go:237]: I think so, I'll make that move.

File: pkg/operator/deploy/deploy.go (hunk 1 above)
Comment (tsatam) [reply pkg/operator/deploy/deploy.go:220]: tl;dr I'll remove this, see below. 


This was done intentionally, but not fully correctly, to keep our two separate Storage Account reconciliation processes, RP-side (cluster creation+PUCM) vs operator, in sync with what they put down. The RP-side reconciliation will include this subnet NACL on just the cluster storageaccount for Hive-created clusters, but the operator would not. 

After re-reading this, it appears that this change as-is will include this NACL on both storageaccounts during operator reconciliation, in reality actually creating the conflict that this change was trying to mitigate (dueling definitions between PUCM and operator).

**Comment by tsatam** [comment]
Closing this for now. We've discussed merging the Subnets and StorageAccounts controllers to allow them to reuse the same ARM GET API calls, and the solution merged in #3280 has been sufficient to reduce the unnecessary reconciles specifically within these two controllers. #3292 (in draft) also has further fixes across the whole operator.

**Pull request closed without merging**